File: backend/routes/exam.py
from flask import Blueprint, request, jsonify

from services.exam_service import ExamService

exam_bp = Blueprint("exam", __name__)


@exam_bp.route("/", methods=["POST"])
def create_exam():

    data = request.get_json()

    required = [
        "academic_year",
        "semester",
        "exam_type",
        "exam_month",
        "exam_year",
        "department_id",
        "created_by",
    ]

    for field in required:

        if field not in data:

            return jsonify({"success": False, "message": f"{field} is required"}), 400

    exam = ExamService.get_or_create(
        academic_year=data["academic_year"],
        semester=data["semester"],
        section=data.get("section"),
        exam_type=data["exam_type"],
        exam_month=data["exam_month"],
        exam_year=data["exam_year"],
        department_id=data["department_id"],
        created_by=data["created_by"],
    )

    return (
        jsonify(
            {
                "success": True,
                "message": "Exam ready",
                "exam": {
                    "id": exam.id,
                    "academic_year": exam.academic_year,
                    "semester": exam.semester,
                    "section": exam.section,
                    "exam_type": exam.exam_type,
                    "exam_month": exam.exam_month,
                    "exam_year": exam.exam_year,
                },
            }
        ),
        201,
    )

File: frontend/lib/models/exam.dart
class Exam {
  final int id;
  final String academicYear;
  final int semester;
  final String? section;
  final String examType;
  final String examMonth;
  final int examYear;

  Exam({
    required this.id,
    required this.academicYear,
    required this.semester,
    this.section,
    required this.examType,
    required this.examMonth,
    required this.examYear,
  });

  factory Exam.fromJson(Map<String, dynamic> json) {
    return Exam(
      id: json["id"],
      academicYear: json["academic_year"],
      semester: json["semester"],
      section: json["section"],
      examType: json["exam_type"],
      examMonth: json["exam_month"],
      examYear: json["exam_year"],
    );
  }

  String get displayName {
    return "Semester $semester • $examType • $examMonth $examYear";
  }
}